### PR TITLE
SONARJAVA-1041 RSPEC-2864 "entrySet()" should be iterated when both the key and value are needed

### PR DESCRIPTION
**Pull request by merciesa**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      return ((IdentifierTree) expressionTree).symbol();
+    } else {
+      return isFieldAccessedUsingSuperOrThis(expressionTree);
+    }
+  }
+
+  @CheckForNull
+  private Symbol isFieldAccessedUsingSuperOrThis(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelectTree = ((MemberSelectExpressionTree) expressionTree);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java]: :new: MAJORRemove those useless parentheses.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      if (result) {
+        addIssue(forEachTree, "Iterate over the \"entrySet\" instead of the \"keySet\".");
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (MAP_GET_METHOD.matches(tree)) {
+        Tree firstArgument = Iterables.getOnlyElement(tree.arguments());
+        if (getMapSymbol(tree).equals(mapSymbol) && firstArgument.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) firstArgument).symbol().equals(variable)) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java]: :new: BLOCKERNullPointerException might be thrown as 'getMapSymbol' is nullable here[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)

File: java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    }
+  }
+
+  @CheckForNull
+  private Symbol isFieldAccessedUsingSuperOrThis(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelectTree = ((MemberSelectExpressionTree) expressionTree);
+      if (memberSelectTree.expression().is(Tree.Kind.IDENTIFIER)) {
+        IdentifierTree identifierTree = (IdentifierTree) memberSelectTree.expression();
+        if (identifierTree.identifierToken().text().equals("super") || identifierTree.identifierToken().text().equals("this")) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java]: :new: MAJORMove the "super" string literal on the left side of this string comparison.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)
:new: MAJORMove the "this" string literal on the left side of this string comparison.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 1 blocking
* 3 major

Details:
* [Remove those useless parentheses.](https://github.com/SonarSource/sonar-java/blob/71a6464e828abfae114b0d3e3e293257c3366fec/java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java#L103)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)
* [Move the "super" string literal on the left side of this string comparison.](https://github.com/SonarSource/sonar-java/blob/71a6464e828abfae114b0d3e3e293257c3366fec/java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java#L106)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)
* [NullPointerException might be thrown as 'getMapSymbol' is nullable here](https://github.com/SonarSource/sonar-java/blob/71a6464e828abfae114b0d3e3e293257c3366fec/java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java#L133)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)
* [Move the "this" string literal on the left side of this string comparison.](https://github.com/SonarSource/sonar-java/blob/71a6464e828abfae114b0d3e3e293257c3366fec/java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java#L106)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S2864.html [hunk 4]
```diff
@@ -0,0 +1,24 @@
+<p>When only the keys from a map are needed in a loop, iterating the <code>keySet</code> makes sense. But when both the key and the value are needed, it's more efficient to iterate the entrySet, which will give access to both the key and value, instead.</p>
```
Comment (m-g-sonar) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S2864.html]: Missing <code> next to the entrySet.

File: java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    ExpressionTree expressionTree = tree.methodSelect();
+    // direct invocation: symbol is implicitly this
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return tree.symbol().owner();
+    }
+    expressionTree = ((MemberSelectExpressionTree) expressionTree).expression();
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return ((IdentifierTree) expressionTree).symbol();
+    } else {
+      return isFieldAccessedUsingSuperOrThis(expressionTree);
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java]: To me, based on its name, the return type of the method should be a Boolean. It should be renamed.

File: java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    }
+  }
+
+  @CheckForNull
+  private Symbol isFieldAccessedUsingSuperOrThis(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelectTree = (MemberSelectExpressionTree) expressionTree;
+      if (memberSelectTree.expression().is(Tree.Kind.IDENTIFIER)) {
+        IdentifierTree identifierTree = (IdentifierTree) memberSelectTree.expression();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java]: why not storing directly the identifierToken ?

File: java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        Symbol mapSymbol = getMapSymbol(methodTree);
+        if (mapSymbol != null && MAP_KEYSET_METHOD.matches(methodTree)) {
+          new GetUsageVisitor().isCallingGetWithSymbol(forEachTree, forEachTree.variable().symbol(), mapSymbol);
+        }
+      }
+    }
+  }
+
+  @CheckForNull
+  private Symbol getMapSymbol(MethodInvocationTree tree) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java]: With this method you are retrieving the symbol corresponding to the owner of the method which is called. As you don't test its type (or it's a map), maybe you should rename the method.

File: java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        addIssue(forEachTree, "Iterate over the \"entrySet\" instead of the \"keySet\".");
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (MAP_GET_METHOD.matches(tree)) {
+        Tree firstArgument = Iterables.getOnlyElement(tree.arguments());
+        if (mapSymbol.equals(getMapSymbol(tree)) && firstArgument.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) firstArgument).symbol().equals(variable)) {
+          result = true;
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/KeySetInsteadOfEntrySetCheck.java]: You can probably nterrupt the exploration at this point with an empty `return;`.

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

**Comment by m-g-sonar** [comment]
LGTM!

**Pull request closed without merging**